### PR TITLE
Allow webauthn to be passed when issuing certs for web-based scp

### PR DESCRIPTION
**Pull request by avatus**
Currently, the web UI cannot download files from the console due to lack of MFA support. After the addition of https://github.com/gravitational/teleport/pull/22528, we can pass a webauthn assertion response to generate certs. This allows the frontend to pass a param that includes their webautn assertion response and if it is present, will issue new certs. If MFA is required on the server and this key is not passed in, they will receive a "mfa required for file transfer" error, rather than a nebulous "key not found" error.

A frontend PR for this is needed and will be PRd again this branch once https://github.com/gravitational/teleport/pull/22529 is merged to avoid conflicts

### CODE REVIEW

File: lib/web/files.go [hunk 1]
```diff
@@ -106,6 +130,10 @@ func (f *fileTransfer) download(req fileTransferRequest, httpReq *http.Request,
 		return trace.Wrap(err)
 	}
 
+	if req.webauthn != "" {
+		f.issueSingleUseCert(req.webauthn, httpReq, tc)
```
Comment (zmb3) [inline lib/web/files.go:134]: Don't forget to check the error here

File: lib/web/files.go [hunk 2]
```diff
@@ -68,6 +76,22 @@ func (h *Handler) transferFile(w http.ResponseWriter, r *http.Request, p httprou
...
+				Login: p.ByName("login"),
+			},
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if mfaReq.Required && query.Get("webauthn") == "" {
+		return nil, trace.BadParameter("MFA required for file transfer.")
```
Comment (zmb3) [inline lib/web/files.go:92]: If MFA is required, and MFA wasn't provided, should this be `trace.AccessDenied` instead?

Comment (avatus) [reply lib/web/files.go]: Yeah that makes sense. My original idea for `BadParameter` was "you didnt send in the parameter that was needed". I think I was in dev mode of "I need to send this parameter" but from the actual user's perspective, it makes more sense to say "You don't have access" . Thanks 👍

**Review by capnspacehook — APPROVED** [review]

File: lib/web/files.go [hunk 3]
```diff
@@ -68,6 +76,22 @@ func (h *Handler) transferFile(w http.ResponseWriter, r *http.Request, p httprou
...
+				Login: p.ByName("login"),
+			},
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if mfaReq.Required && query.Get("webauthn") == "" {
+		return nil, trace.AccessDenied("MFA required for file transfer.")
```
Comment (zmb3) [inline lib/web/files.go:92]:
```suggestion
		return nil, trace.AccessDenied("MFA required for file transfer")
```

File: lib/web/files.go [hunk 4]
```diff
@@ -179,3 +217,46 @@ func (f *fileTransfer) createClient(req fileTransferRequest, httpReq *http.Reque
...
+
+	key, err := client.GenerateRSAKey()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cert, err := f.authClient.GenerateUserCerts(httpReq.Context(), proto.UserCertsRequest{
+		PublicKey: key.MarshalSSHPublicKey(),
+		Username:  f.ctx.GetUser(),
+		Expires:   time.Now().Add(time.Minute).UTC(),
```
Comment (zmb3) [inline lib/web/files.go:241]: Is 1 minute enough? Should we add a little buffer to account for clock drift, or is 1m what we use elsewhere too?

File: lib/web/files.go [hunk 5]
```diff
@@ -130,6 +161,13 @@ func (f *fileTransfer) upload(req fileTransferRequest, httpReq *http.Request) er
 		return trace.Wrap(err)
 	}
 
+	if req.webauthn != "" {
+		err = f.issueSingleUseCert(req.webauthn, httpReq, tc)
```
Comment (zmb3) [inline lib/web/files.go:165]: At first it wasn't clear to me what happens to the single use cert that gets issued. Maybe a godoc on `issueSingleUseCert` that explains that it configures `tc` to use the new cert would help.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**